### PR TITLE
SUBMARINE-361. Run submarine operator in k8s

### PR DESCRIPTION
**Pull request by liuxunorg**
### What is this PR for?
Run the submarine operator image in k8s as a pod. Provide the service of the submarine server.
1. Create submarine cluster CRD automatically
2. Support creating submarine server in k8s, detecting submarine server running status, providing high availability

### What type of PR is it?
[Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/SUBMARINE/
* Put link here, and add [SUBMARINE-*Jira number*] in PR title, eg. [SUBMARINE-23]

### How should this be tested?
* https://travis-ci.org/liuxunorg/submarine/builds/642231831

### Screenshots (if appropriate)

image
image

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: submarine-cloud/pkg/utils/build.go [hunk 1]
```diff
@@ -23,25 +23,30 @@ import (
 
 // BUILDTIME should be populated by at build time: -ldflags "-w -X github.com/apache/submarine/submarine-cloud/pkg/utils.BUILDTIME=${DATE}
 // with for example DATE=$(shell date +%Y-%m-%d/%H:%M:%S )   (pay attention not to use space!)
-var BUILDTIME string
+var BuildTime string
 
-// TAG should be populated by at build time: -ldflags "-w -X github.com/apache/submarine/submarine-cloud/pkg/utils.TAG=${TAG}
-// with for example TAG=$(shell git tag|tail -1)
-var TAG string
+// TAG should be populated by at build time: -ldflags "-w -X github.com/apache/submarine/submarine-cloud/pkg/utils.TAG=${BuildGitBranch}
```
Comment (yuanzac) [inline submarine-cloud/pkg/utils/build.go]: The comment is about TAG?

File: submarine-cloud/pkg/utils/build.go [hunk 2]
```diff
@@ -23,25 +23,30 @@ import (
...
+// with for example BuildGitBranch=$(git describe --all)
+var BuildGitBranch string
+
+// TAG should be populated by at build time: -ldflags "-w -X github.com/apache/submarine/submarine-cloud/pkg/utils.TAG=${BuildGitRev}
```
Comment (yuanzac) [inline submarine-cloud/pkg/utils/build.go]: The comment is about TAG?

File: submarine-cloud/manifests/crd.yaml [hunk 3]
```diff
@@ -1,45 +1,53 @@
-#
```
[the remark this answers is not in the document]
Comment (liuxunorg) [reply submarine-cloud/manifests/crd.yaml]: Done.

File: submarine-cloud/pkg/controller/actions.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+		}
+	}
+
+	clusterStatus := &cluster.Status.Cluster
+	if (clusterStatus.NbPods - clusterStatus.NbSubmarineRunning) != 0 {
+		glog.V(3).Infof("All pods not ready wait to be ready, nbPods: %d, nbPodsReady: %d", clusterStatus.NbPods, clusterStatus.NbSubmarineRunning)
+		return false, err
+	}
+
+	//First, we define the new masters
```
[the remark this answers is not in the document]
Comment (liuxunorg) [reply submarine-cloud/pkg/controller/actions.go]: Done.

File: submarine-cloud/pkg/utils/build.go (hunk 1 above)
Comment (liuxunorg) [reply submarine-cloud/pkg/utils/build.go]: Done.

File: submarine-cloud/pkg/utils/build.go (hunk 2 above)
Comment (liuxunorg) [reply submarine-cloud/pkg/utils/build.go]: Done.

File: submarine-cloud/pkg/controller/actions.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+	"github.com/apache/submarine/submarine-cloud/pkg/submarine"
+	"github.com/golang/glog"
+	"time"
+)
+
+// Perform various management operations on Submarine Pod and Submarine clusters to approximate the desired state
+func (c *Controller) clusterAction(admin submarine.AdminInterface, cluster *rapi.SubmarineCluster, infos *submarine.ClusterInfos) (bool, error) {
+	glog.Info("clusterAction()")
+	var err error
+	/* run sanity check if needed
```
[the remark this answers is not in the document]
Comment (liuxunorg) [reply submarine-cloud/pkg/controller/actions.go]: I will continue to submit subsequent PRs to complete this work.

**Pull request closed without merging**